### DISCUSSION TITLE
Reasons for !Send.

### DISCUSSION THREAD

**Discussion by navneetankur** [mlua-rs/mlua#398] [Q&A]
mlua::Lua is !Send by default. What is the reason for it except for existence of Owned types like OwnedTable?

Internally I see it uses Rc. But it should be fine if nothing Rc-like can be taken out of mlua::Lua correct?
And what is difference between with and without [send] feature. Other than Rc/Arc and Owned types?

Basically If I wrap mlua::Lua in [Exclusive](https://doc.rust-lang.org/nightly/std/sync/struct.Exclusive.html), can I forcefully mark it as Send through newtype?

**Comment by khvzak** [comment]
> mlua::Lua is !Send by default. What is the reason for it except for existence of Owned types like OwnedTable?
> And what is difference between with and without [send] feature. Other than Rc/Arc and Owned types?

`send` feature flag adds `Send` requirements for Rust callbacks and userdata types stored in Lua via `MaybeSend` trait.
Given that Lua itself is single threaded, not having `Send` requirement by default makes a lot of sense.

> Internally I see it uses Rc. But it should be fine if nothing Rc-like can be taken out of mlua::Lua correct?

Could you point please where `Rc` (not `Arc`) is used?

**Reply by navneetankur** [reply]
> send feature flag adds Send requirements for Rust callbacks and userdata types stored in Lua via MaybeSend trait.

Okay, thanks. I was wondering if it does something else other than marking it Send and disabling owned type returning fn.

> Given that Lua itself is single threaded, not having Send requirement by default makes a lot of sense.

I don't understand this statement completely. From purely theoretical perspective single threaded means it should not be `Sync`. It can still be send like `RefCell` if it does not make a shallow clone which can be left behind in the current thread, if sent to another thread (the way `Rc` does). Of course there can be other non-general reason for something to not be `Send` like per thread platform specific setup before running or thread local storage or something.

> Could you point please where Rc (not Arc) is used?

When I first asked this question a think I saw some logic which uses Rc if #[cfg(not(feature = "send"))]. Which would make sense, but made me wonder if !Send only due to use of Rc.
[TypeId::of::<Rc<RefCell<T>>>()](https://github.com/mlua-rs/mlua/blob/8f3de8aa19ac7263297b080862c0a7a0ddb5de04/src/userdata_impl.rs#L99).
